libkbfs/tlf_edit_history.go

```diff
@@ -0,0 +1,355 @@
+
+				writer := op.getWriterInfo().uid
+				createdPath := op.getFinalPath().ChildPathNoPtr(realOp.NewName)
+				edits[writer] = append(edits[writer], TlfEdit{
+					Filepath:  createdPath.String(),
+					Type:      FileCreated,
+					LocalTime: op.getLocalTimestamp(),
+				})
+			case *syncOp:
+				lastOp := op
+				// Only the final writer matters, so find the last
+				// syncOp in this chain.
+				for j := len(chain.ops) - 1; j > i; j-- {
+					if syncOp, ok := chain.ops[j].(*syncOp); ok {
+						lastOp = syncOp
+						break
+					}
+				}
+
+				writer := lastOp.getWriterInfo().uid
+				t := FileModified
+				if chains.isCreated(ptr) {
+					t = FileCreated
+				}
+				edits[writer] = append(edits[writer], TlfEdit{
+					Filepath:  lastOp.getFinalPath().String(),
+					Type:      t,
+					LocalTime: lastOp.getLocalTimestamp(),
+				})
+				continue outer
+			default:
+				continue
+			}
+		}
+	}
+
+	return edits, nil
+}
+
+// GetComplete returns the most recently known set of clustered edit
+// history for this TLF.
+func (teh *TlfEditHistory) GetComplete(ctx context.Context,
+	head ImmutableRootMetadata) (TlfWriterEdits, error) {
+	var currEdits TlfWriterEdits
+	/**
+	* Once we update currEdits based on notifications, we can uncomment this.
+		currEdits := teh.getEditsCopy()
+		if currEdits != nil {
+			return currEdits, nil
+		}
+	*/
+
+	// We have no history -- fetch from the server until we have a
+	// complete history.
+
+	estimates := make(writerEditEstimates)
+	for _, writer := range head.GetTlfHandle().ResolvedWriters() {
+		estimates[writer] = 0
+	}
+	rmds := []ImmutableRootMetadata{head}
+	estimates.update(rmds)
+
+	// If unmerged, get all the unmerged updates.
+	if head.MergedStatus() == Unmerged {
+		_, unmergedRmds, err := getUnmergedMDUpdates(ctx, teh.config, head.ID,
+			head.BID, head.Revision-1)
+		if err != nil {
+			return nil, err
+		}
+		estimates.update(unmergedRmds)
+		rmds, err = teh.updateRmds(rmds, unmergedRmds)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for (currEdits == nil || !currEdits.isComplete()) &&
+		len(rmds) < maxMDsToInspect &&
+		rmds[0].Revision > MetadataRevisionInitial {
+		teh.log.CDebugf(ctx, "Edits not complete after %d revisions", len(rmds))
+		if estimates.isComplete() {
+			// Once the estimate hits the threshold for each writer,
+			// calculate the chains using all those MDs, and build the
+			// real edit map (discounting deleted files, etc).
+			var err error
+			currEdits, err = teh.calculateEditCounts(ctx, rmds)
+			if err != nil {
+				return nil, err
+			}
+			if currEdits.isComplete() {
+				break
+			}
+
+			// Set the estimates to their exact known values
+			estimates.reset(currEdits)
+		}
+
+		for !estimates.isComplete() && len(rmds) < maxMDsToInspect &&
+			rmds[0].Revision > MetadataRevisionInitial {
+			// Starting from the head/branchpoint, work backwards
+			// mdMax revisions at a time.
+			endRev := rmds[0].Revision - 1
+			startRev := endRev - maxMDsAtATime + 1
```

Comment:
I don't care that much, but now that you've pointed it out it bugs me.  Will fix.